torch/optim/decoupled_weight_decay_optim.py

```diff
@@ -0,0 +1,99 @@
+        scheduler is used for the learning rate, it is reasonable to schedule
+        the weight decay as well.
+
+    .. note:: To prevent copying of the parameters, the weight decay is
+        applied before the optimizer update. If the update step relies on
+        the value of the parameter (and not just the gradient), this function
+        should not be used.
+
+    Example:
+        >>> AdamW = extend_with_decoupled_weight_decay(torch.optim.Adam)
+        >>> optimizer = AdamW(decoupled_weight_decay=1e-5, lr=1e-3)
+    """
+    
+    class DecoupledWeightDecayOptimizer(base_optimizer):
+
+        __doc__ = \
+        """
+        Implements the {0}W algorithm.
+            
+        I.e., this class implements the update step of {0} with weight
+        decay, where the weight decay is decoupled from the optimization
+        steps w.r.t. to the loss function, as proposed in Decoupled Weight
+        Decay Regularization by  Loshchilov & Hutter:
+        https://arxiv.org/abs/1711.05101
+
+        .. note:: The weight decay is not coupled with the  learning rate.
+            If a scheduler is used for the learning  rate, it is reasonable to
+            schedule the weight decay as well.
+        
+        Arguments:
+            decoupled_weight_decay (float, optional): weight decay (default: 0)
+            *args: arguments passed to the base optimizer.
+            **kwargs: keyword arguments passed to the base optimizer.
+        
+        Details on the update step can be found in the documentation of the
+        wrapped {0} optimizer.
+        """.format(base_optimizer.__name__)
+
+        def __init__(self, decoupled_weight_decay, *args, **kwargs):
+            super(DecoupledWeightDecayOptimizer, self).__init__(*args, **kwargs)
+            if self.defaults["weight_decay"] != 0:
+                warnings.warn(
+                    "The weight decay parameter of the base optimizer is not "
+                    "0. This means the weights will be decayed with l2 loss "
+                    "AND decoupled weight decay.", UserWarning)
+            self.defaults["decoupled_weight_decay"] = decoupled_weight_decay
+            # update all param groups to contain the default weight decay
+            # by re-adding them
+            for group in self.param_groups:
+                group.setdefault("decoupled_weight_decay", decoupled_weight_decay)
+
+        def step(self, closure=None):
+            # First perform the weight decay, then execute the optimizer step.
+            # This way, we don't need to store the parameter value prior to the
+            # update. The drawback is, that the optimizer's step function must
+            # not depend on the parameter value but only the gradient.
+            for group in self.param_groups:
+                for p in group['params']:
+                    if p.grad is not None:
+                        p.data.mul_(1 - group['decoupled_weight_decay'])
```

Comment:
We'll need tests here showing under which condition this implementation and the current one agree.

Reply:
This I haven't included yet, what's the format you suggest? Keep the old optimizer and compare the outputs?